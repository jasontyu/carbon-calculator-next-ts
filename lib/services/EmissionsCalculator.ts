// to unwrap ctype from input without using them
/* eslint-disable @typescript-eslint/no-unused-vars */

import { CalculateApi } from "../../pages/api/calculate"

export type CalculationResult = {
  emissions: number
}

// Intermediary type with type discrimination allows mapping of generic calculate to strongly typed helpers
export type CalculationInput = { ctype: 'food' } & CalculateApi.RequestBody['calculations']['food']
  | { ctype: 'transportation' } & CalculateApi.RequestBody['calculations']['transportation']

export const calculate = (input: CalculationInput): CalculationResult => {
  switch (input.ctype) {
    case 'food': {
      const { ctype, ...data } = input
      return calculateFood(data)
    }
    case 'transportation': {
      const { ctype, ...data } = input
      return calculateTransportation(data)
    }
  }
}

// TODO: extract request body schema so we can separate the .partial() which is contaminating internal types
// That way, we won't have to manually wrap types in Required<..>
export const calculateFood = (data: Required<CalculateApi.RequestBody['calculations']>['food']) => {
  // Median GHG emissions (kg CO2eq / NU)
  // from aaq0216_datas2 (https://www.science.org/doi/10.1126/science.aaq0216)
  const factors: Record<keyof typeof data, number> = {
    bread: 0.5 / 1000,
    meat: 30 / 1000,
    vegetables: 0.4 / 1000
  }
  const emissions = (Object.keys(data) as (keyof typeof data)[]) // TS loses type safety when iterating objects so we need to do this redundant looking cast
    .map(key => factors[key] * data[key])
    .reduce((sum, next) => sum + next, 0)

  return { emissions }
}

export const calculateTransportation = (data: Required<CalculateApi.RequestBody['calculations']>['transportation']) => {
  // CO2 Factor (kg CO2 / vehicle-mile)
  // from Emission Factors for Greenhouse Gas Inventories
  // Table 10: Scope 3 Category 6: Business Travel and Category 7: Employee Commuting
  const factors: Record<keyof typeof data, number> = {
    plane: 0.133, // Air Travel, medium-haul (passenger-mile)
    car: 0.335, // Passenger Car
    bus: 0.053 // Bus
  }
  const emissions = (Object.keys(data) as (keyof typeof data)[]) // TS loses type safety when iterating objects so we need to do this redundant looking cast
    .map(key => factors[key] * data[key])
    .reduce((sum, next) => sum + next, 0)

  return { emissions }
}
