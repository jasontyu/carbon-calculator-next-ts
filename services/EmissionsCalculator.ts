import { CalculationType } from "../pages/api/calculate"

export type CalculationResult = {
  emissions: number
}

// TODO: Pass through calculation (with type safety if possible)
// TODO: Write unit tests for Calculator functions
export const calculate = (type: CalculationType, inputs: unknown): CalculationResult=> {
  const mapping: Record<CalculationType, (data: unknown) => CalculationResult> = {
    transportation: calculateTransportation,
    food: calculateFood
  }
  return mapping[type](inputs)
}

const calculateTransportation = (inputs: unknown) => ({ emissions: 42 })
const calculateFood = (inputs: unknown) => ({ emissions: 9999 })