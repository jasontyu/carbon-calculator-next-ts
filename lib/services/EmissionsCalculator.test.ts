import 'jest'
import * as EmissionsCalculator from './EmissionsCalculator'

describe('EmissionsCalculator', () => {

  describe('`calculate` calls through to helper and matches snapshot', () => {
    type TestCase = { input: EmissionsCalculator.CalculationInput, helper: keyof typeof EmissionsCalculator }
    const cases: TestCase[] = [
      { input: { ctype: 'food', bread: 100, meat: 200, vegetables: 300 }, helper: 'calculateFood' },
      { input: { ctype: 'transportation', bus: 100, car: 200, plane: 300 }, helper: 'calculateTransportation' },
    ]

    // This is the more intuitive approach, but it doesn't work
    // because the mapping in `EmissionsCalculator.calculate` does not get spied on properly
    // const itShouldCallThroughCorrectly = (testCase: TestCase) => {
    //     const { input, spy } = testCase
    //     it(`for type '${input.ctype}'`, () => {
    //     const { ctype, ...data } = input
    //       const spyInstance = jest.spyOn(EmissionsCalculator, spy)
    //       EmissionsCalculator.calculate(input)
    //       expect(spyInstance).toHaveBeenCalledWith(data)
    //   })
    // }
    // cases.forEach(testCase => itShouldCallThroughCorrectly(testCase))

    // So instead, we use a slightly hackier way to assert that it calls through correctly
    const itShouldCallThroughAndMatchSnapshot = (testCase: TestCase) => {
      const { input, helper } = testCase
      it(`for type '${input.ctype}'`, () => {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const { ctype, ...data } = input
        const result1 = EmissionsCalculator.calculate(input)
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const result2 = (EmissionsCalculator[helper] as any)(data)
        expect(result1).toStrictEqual(result2)
        expect(result1).toMatchSnapshot()
      })
    }
    cases.forEach(testCase => itShouldCallThroughAndMatchSnapshot(testCase))
  })
})