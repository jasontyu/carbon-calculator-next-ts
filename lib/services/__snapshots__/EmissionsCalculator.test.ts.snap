// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmissionsCalculator \`calculate\` calls through to helper and matches snapshot for type 'food' 1`] = `
Object {
  "emissions": 6.17,
}
`;

exports[`EmissionsCalculator \`calculate\` calls through to helper and matches snapshot for type 'transportation' 1`] = `
Object {
  "emissions": 112.2,
}
`;
